source/common/singleton/manager_impl.cc

```diff
@@ -0,0 +1,26 @@
+#include "common/singleton/manager_impl.h"
+
+#include "envoy/registry/registry.h"
+
+namespace Envoy {
+namespace Singleton {
+
+InstancePtr ManagerImpl::get(const std::string& name) {
+  verifyRegistration(name);
+  return singletons_[name].lock();
+}
+
+void ManagerImpl::set(const std::string& name, InstancePtr singleton) {
+  verifyRegistration(name);
+  singletons_[name] = singleton;
+}
+
+void ManagerImpl::verifyRegistration(const std::string& name) {
+  if (nullptr == Registry::FactoryRegistry<Registration>::getFactory(name)) {
```

Comment:
We'd probably catch most of the core Envoy violations of this in the CI runs if it was an ASSERT, but I see the value of making it an exception or panic when Envoy is extended by other folks. So, up to you what you feel is best on this, just wanted to clarify my understanding.